<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.8.3.1"/>
<title>Qualia: src/qualia/ga/Chromosome.h Source File</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td style="padding-left: 0.5em;">
   <div id="projectname">Qualia
   &#160;<span id="projectnumber">0.2</span>
   </div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.8.3.1 -->
  <div id="navrow1" class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&#160;Page</span></a></li>
      <li><a href="annotated.html"><span>Classes</span></a></li>
      <li class="current"><a href="files.html"><span>Files</span></a></li>
    </ul>
  </div>
  <div id="navrow2" class="tabs2">
    <ul class="tablist">
      <li><a href="files.html"><span>File&#160;List</span></a></li>
      <li><a href="globals.html"><span>File&#160;Members</span></a></li>
    </ul>
  </div>
<div id="nav-path" class="navpath">
  <ul>
<li class="navelem"><a class="el" href="dir_68267d1309a1af8e8297ef4c3efbcdba.html">src</a></li><li class="navelem"><a class="el" href="dir_0668cf3d5adc15b485aff550958bf699.html">qualia</a></li><li class="navelem"><a class="el" href="dir_773568f85717bfb4dd35c49bec140958.html">ga</a></li>  </ul>
</div>
</div><!-- top -->
<div class="header">
  <div class="headertitle">
<div class="title">Chromosome.h</div>  </div>
</div><!--header-->
<div class="contents">
<a href="Chromosome_8h.html">Go to the documentation of this file.</a><div class="fragment"><div class="line"><a name="l00001"></a><span class="lineno">    1</span>&#160;<span class="comment">/*</span></div>
<div class="line"><a name="l00002"></a><span class="lineno">    2</span>&#160;<span class="comment"> * Chromosome.h</span></div>
<div class="line"><a name="l00003"></a><span class="lineno">    3</span>&#160;<span class="comment"> *</span></div>
<div class="line"><a name="l00004"></a><span class="lineno">    4</span>&#160;<span class="comment"> * (c) 2012 Sofian Audry -- info(@)sofianaudry(.)com</span></div>
<div class="line"><a name="l00005"></a><span class="lineno">    5</span>&#160;<span class="comment"> *</span></div>
<div class="line"><a name="l00006"></a><span class="lineno">    6</span>&#160;<span class="comment"> * This program is free software: you can redistribute it and/or modify</span></div>
<div class="line"><a name="l00007"></a><span class="lineno">    7</span>&#160;<span class="comment"> * it under the terms of the GNU General Public License as published by</span></div>
<div class="line"><a name="l00008"></a><span class="lineno">    8</span>&#160;<span class="comment"> * the Free Software Foundation, either version 3 of the License, or</span></div>
<div class="line"><a name="l00009"></a><span class="lineno">    9</span>&#160;<span class="comment"> * (at your option) any later version.</span></div>
<div class="line"><a name="l00010"></a><span class="lineno">   10</span>&#160;<span class="comment"> *</span></div>
<div class="line"><a name="l00011"></a><span class="lineno">   11</span>&#160;<span class="comment"> * This program is distributed in the hope that it will be useful,</span></div>
<div class="line"><a name="l00012"></a><span class="lineno">   12</span>&#160;<span class="comment"> * but WITHOUT ANY WARRANTY; without even the implied warranty of</span></div>
<div class="line"><a name="l00013"></a><span class="lineno">   13</span>&#160;<span class="comment"> * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the</span></div>
<div class="line"><a name="l00014"></a><span class="lineno">   14</span>&#160;<span class="comment"> * GNU General Public License for more details.</span></div>
<div class="line"><a name="l00015"></a><span class="lineno">   15</span>&#160;<span class="comment"> *</span></div>
<div class="line"><a name="l00016"></a><span class="lineno">   16</span>&#160;<span class="comment"> * You should have received a copy of the GNU General Public License</span></div>
<div class="line"><a name="l00017"></a><span class="lineno">   17</span>&#160;<span class="comment"> * along with this program.  If not, see &lt;http://www.gnu.org/licenses/&gt;.</span></div>
<div class="line"><a name="l00018"></a><span class="lineno">   18</span>&#160;<span class="comment"> */</span></div>
<div class="line"><a name="l00019"></a><span class="lineno">   19</span>&#160;</div>
<div class="line"><a name="l00020"></a><span class="lineno">   20</span>&#160;<span class="preprocessor">#ifndef CHROMOSOME_H_</span></div>
<div class="line"><a name="l00021"></a><span class="lineno">   21</span>&#160;<span class="preprocessor"></span><span class="preprocessor">#define CHROMOSOME_H_</span></div>
<div class="line"><a name="l00022"></a><span class="lineno">   22</span>&#160;<span class="preprocessor"></span></div>
<div class="line"><a name="l00023"></a><span class="lineno">   23</span>&#160;<span class="keyword">class </span><a class="code" href="classChromosome.html">Chromosome</a>;</div>
<div class="line"><a name="l00024"></a><span class="lineno">   24</span>&#160;</div>
<div class="line"><a name="l00025"></a><span class="lineno">   25</span>&#160;<span class="comment">/*</span></div>
<div class="line"><a name="l00026"></a><span class="lineno">   26</span>&#160;<span class="comment"> NOTE: This section was directly copied from file GAGenome.h in GAlib</span></div>
<div class="line"><a name="l00027"></a><span class="lineno">   27</span>&#160;<span class="comment">       Copyright Massachusetts Institute of Technology and Matthew Wall.</span></div>
<div class="line"><a name="l00028"></a><span class="lineno">   28</span>&#160;<span class="comment"> TODO: - adapt the text to the Qualia GA API</span></div>
<div class="line"><a name="l00029"></a><span class="lineno">   29</span>&#160;<span class="comment"></span></div>
<div class="line"><a name="l00030"></a><span class="lineno">   30</span>&#160;<span class="comment">initialize</span></div>
<div class="line"><a name="l00031"></a><span class="lineno">   31</span>&#160;<span class="comment">  Use this method to set the initial state of your genomes once they have</span></div>
<div class="line"><a name="l00032"></a><span class="lineno">   32</span>&#160;<span class="comment">  been created.  This initialization is for setting up the genome&#39;s state,</span></div>
<div class="line"><a name="l00033"></a><span class="lineno">   33</span>&#160;<span class="comment">  not for doing the basic mechanics of genome class management.  The</span></div>
<div class="line"><a name="l00034"></a><span class="lineno">   34</span>&#160;<span class="comment">  default behaviour of this method is to change randomly the contents of the</span></div>
<div class="line"><a name="l00035"></a><span class="lineno">   35</span>&#160;<span class="comment">  genome.  If you want to bias your initial population, this is where to</span></div>
<div class="line"><a name="l00036"></a><span class="lineno">   36</span>&#160;<span class="comment">  make that happen.</span></div>
<div class="line"><a name="l00037"></a><span class="lineno">   37</span>&#160;<span class="comment">     The initializer is used to initialize the genome (duh).  Notice that the</span></div>
<div class="line"><a name="l00038"></a><span class="lineno">   38</span>&#160;<span class="comment">  state of the genome is unknown - memory may or may not have been allocated,</span></div>
<div class="line"><a name="l00039"></a><span class="lineno">   39</span>&#160;<span class="comment">  and the genome may or may not have been used before.  So your initializer</span></div>
<div class="line"><a name="l00040"></a><span class="lineno">   40</span>&#160;<span class="comment">  should first clean up as needed, then do its thing.  The initializer may be</span></div>
<div class="line"><a name="l00041"></a><span class="lineno">   41</span>&#160;<span class="comment">  called any number of times (unlike a class constructor which is called only</span></div>
<div class="line"><a name="l00042"></a><span class="lineno">   42</span>&#160;<span class="comment">  once for a given instance).</span></div>
<div class="line"><a name="l00043"></a><span class="lineno">   43</span>&#160;<span class="comment"></span></div>
<div class="line"><a name="l00044"></a><span class="lineno">   44</span>&#160;<span class="comment">mutate</span></div>
<div class="line"><a name="l00045"></a><span class="lineno">   45</span>&#160;<span class="comment">  Mutate the genome with probability as specified.  What mutation means</span></div>
<div class="line"><a name="l00046"></a><span class="lineno">   46</span>&#160;<span class="comment">  depends upon the data type of the genome.  For example, you could have</span></div>
<div class="line"><a name="l00047"></a><span class="lineno">   47</span>&#160;<span class="comment">  a bit string in which 50% mutation means each bit has a 50% chance of</span></div>
<div class="line"><a name="l00048"></a><span class="lineno">   48</span>&#160;<span class="comment">  getting flipped, or you could have a tree in which 50% mutation means each</span></div>
<div class="line"><a name="l00049"></a><span class="lineno">   49</span>&#160;<span class="comment">  node has a 50% chance of getting deleted, or you could have a bit string</span></div>
<div class="line"><a name="l00050"></a><span class="lineno">   50</span>&#160;<span class="comment">  in which 50% mutation means 50% of the bits ACTUALLY get flipped.</span></div>
<div class="line"><a name="l00051"></a><span class="lineno">   51</span>&#160;<span class="comment">    The mutations member returns the number of mutations since the genome</span></div>
<div class="line"><a name="l00052"></a><span class="lineno">   52</span>&#160;<span class="comment">  was initialized.</span></div>
<div class="line"><a name="l00053"></a><span class="lineno">   53</span>&#160;<span class="comment">    The mutator makes a change to the genome with likeliehood determined by the</span></div>
<div class="line"><a name="l00054"></a><span class="lineno">   54</span>&#160;<span class="comment">  mutation rate parameter.  The exact meaning of mutation is up to you, as is</span></div>
<div class="line"><a name="l00055"></a><span class="lineno">   55</span>&#160;<span class="comment">  the specific meaning of the mutation rate.  The function returns the number</span></div>
<div class="line"><a name="l00056"></a><span class="lineno">   56</span>&#160;<span class="comment">  of mutations that actually occurred.</span></div>
<div class="line"><a name="l00057"></a><span class="lineno">   57</span>&#160;<span class="comment"></span></div>
<div class="line"><a name="l00058"></a><span class="lineno">   58</span>&#160;<span class="comment">crossover</span></div>
<div class="line"><a name="l00059"></a><span class="lineno">   59</span>&#160;<span class="comment">  Genomes don&#39;t really have any clue about other genomes, so we don&#39;t make</span></div>
<div class="line"><a name="l00060"></a><span class="lineno">   60</span>&#160;<span class="comment">  the crossover a member function.  Instead, each genome kind of knows how</span></div>
<div class="line"><a name="l00061"></a><span class="lineno">   61</span>&#160;<span class="comment">  to mate with other genomes to generate offspring, but they are not</span></div>
<div class="line"><a name="l00062"></a><span class="lineno">   62</span>&#160;<span class="comment">  capable of doing it themselves.  The crossover member function is used to</span></div>
<div class="line"><a name="l00063"></a><span class="lineno">   63</span>&#160;<span class="comment">  set the default mating mode for the genomes - it does not actually perform</span></div>
<div class="line"><a name="l00064"></a><span class="lineno">   64</span>&#160;<span class="comment">  the crossover.  This way the GA can use asexual crossover if it wants to</span></div>
<div class="line"><a name="l00065"></a><span class="lineno">   65</span>&#160;<span class="comment">  (but genomes only know how to do the default sexual crossover).</span></div>
<div class="line"><a name="l00066"></a><span class="lineno">   66</span>&#160;<span class="comment">    This also lets you do funky stuff like crossover between different data</span></div>
<div class="line"><a name="l00067"></a><span class="lineno">   67</span>&#160;<span class="comment">  types and group sex to generate new offspring.</span></div>
<div class="line"><a name="l00068"></a><span class="lineno">   68</span>&#160;<span class="comment">     We define two types of crossover:  sexual and asexual.  Most GAlib</span></div>
<div class="line"><a name="l00069"></a><span class="lineno">   69</span>&#160;<span class="comment">  algorithms use the sexual crossover, but both are available.  Each genome</span></div>
<div class="line"><a name="l00070"></a><span class="lineno">   70</span>&#160;<span class="comment">  knows the preferred crossover method, but none is capable of reproducing.</span></div>
<div class="line"><a name="l00071"></a><span class="lineno">   71</span>&#160;<span class="comment">  The genetic algorithm must actually perform the mating because it involves</span></div>
<div class="line"><a name="l00072"></a><span class="lineno">   72</span>&#160;<span class="comment">  another genome (as parent and/or child).</span></div>
<div class="line"><a name="l00073"></a><span class="lineno">   73</span>&#160;<span class="comment"></span></div>
<div class="line"><a name="l00074"></a><span class="lineno">   74</span>&#160;<span class="comment">evaluator</span></div>
<div class="line"><a name="l00075"></a><span class="lineno">   75</span>&#160;<span class="comment">  Set the genome&#39;s objective function.  This also sets marks the evaluated</span></div>
<div class="line"><a name="l00076"></a><span class="lineno">   76</span>&#160;<span class="comment">  flag to indicate that the genome must be re-evaluated.</span></div>
<div class="line"><a name="l00077"></a><span class="lineno">   77</span>&#160;<span class="comment">    Evaluation happens on-demand - the objective score is not calculated until</span></div>
<div class="line"><a name="l00078"></a><span class="lineno">   78</span>&#160;<span class="comment">  it is requested.  Then it is cached so that it does not need to be re-</span></div>
<div class="line"><a name="l00079"></a><span class="lineno">   79</span>&#160;<span class="comment">  calculated each time it is requested.  This means that any member function</span></div>
<div class="line"><a name="l00080"></a><span class="lineno">   80</span>&#160;<span class="comment">  that modifies the state of the genome must also set the evaluated flag to</span></div>
<div class="line"><a name="l00081"></a><span class="lineno">   81</span>&#160;<span class="comment">  indicate that the score must be recalculated.</span></div>
<div class="line"><a name="l00082"></a><span class="lineno">   82</span>&#160;<span class="comment">    The genome objective function is used by the GA to evaluate each member of</span></div>
<div class="line"><a name="l00083"></a><span class="lineno">   83</span>&#160;<span class="comment">  the population.</span></div>
<div class="line"><a name="l00084"></a><span class="lineno">   84</span>&#160;<span class="comment"></span></div>
<div class="line"><a name="l00085"></a><span class="lineno">   85</span>&#160;<span class="comment">comparator</span></div>
<div class="line"><a name="l00086"></a><span class="lineno">   86</span>&#160;<span class="comment">  This method is used to determine how similar two genomes are.  If you want</span></div>
<div class="line"><a name="l00087"></a><span class="lineno">   87</span>&#160;<span class="comment">  to use a different comparison method without deriving a new class, then use</span></div>
<div class="line"><a name="l00088"></a><span class="lineno">   88</span>&#160;<span class="comment">  the comparator function to do so.  For example, you may want to do phenotype-</span></div>
<div class="line"><a name="l00089"></a><span class="lineno">   89</span>&#160;<span class="comment">  based comparisons rather than genotype-based comparisons.</span></div>
<div class="line"><a name="l00090"></a><span class="lineno">   90</span>&#160;<span class="comment">    In many cases we have to compare two genomes to determine how similar or</span></div>
<div class="line"><a name="l00091"></a><span class="lineno">   91</span>&#160;<span class="comment">  different they are.  In traditional GA literature this type of function is</span></div>
<div class="line"><a name="l00092"></a><span class="lineno">   92</span>&#160;<span class="comment">  referred to as a &#39;distance&#39; function, probably because bit strings can be</span></div>
<div class="line"><a name="l00093"></a><span class="lineno">   93</span>&#160;<span class="comment">  compared using the Hamming distance as a measure of similarity.  In GAlib, we</span></div>
<div class="line"><a name="l00094"></a><span class="lineno">   94</span>&#160;<span class="comment">  define a genome comparator function that does exactly this kind of</span></div>
<div class="line"><a name="l00095"></a><span class="lineno">   95</span>&#160;<span class="comment">  comparison.</span></div>
<div class="line"><a name="l00096"></a><span class="lineno">   96</span>&#160;<span class="comment">    If the genomes are identical, the similarity function should return a</span></div>
<div class="line"><a name="l00097"></a><span class="lineno">   97</span>&#160;<span class="comment">  value of 0.0, if completely different then return a value greater than 0.</span></div>
<div class="line"><a name="l00098"></a><span class="lineno">   98</span>&#160;<span class="comment">  The specific definition of what &quot;the same&quot; and what &quot;different&quot; mean is up</span></div>
<div class="line"><a name="l00099"></a><span class="lineno">   99</span>&#160;<span class="comment">  to you.  Most of the default comparators use the genotype for the comparison,</span></div>
<div class="line"><a name="l00100"></a><span class="lineno">  100</span>&#160;<span class="comment">  but you can use the phenotype if you prefer.  There is no upper limit to the</span></div>
<div class="line"><a name="l00101"></a><span class="lineno">  101</span>&#160;<span class="comment">  distance score as far as GAlib is concerned.</span></div>
<div class="line"><a name="l00102"></a><span class="lineno">  102</span>&#160;<span class="comment">    The no-op function returns a -1 to signify that the comparison failed.</span></div>
<div class="line"><a name="l00103"></a><span class="lineno">  103</span>&#160;<span class="comment"></span></div>
<div class="line"><a name="l00104"></a><span class="lineno">  104</span>&#160;<span class="comment">//evalData</span></div>
<div class="line"><a name="l00105"></a><span class="lineno">  105</span>&#160;<span class="comment">//  The evalData member is useful if you do not want to derive a new genome class</span></div>
<div class="line"><a name="l00106"></a><span class="lineno">  106</span>&#160;<span class="comment">//  but want to store data with each genome.  When you clone a genome, the eval</span></div>
<div class="line"><a name="l00107"></a><span class="lineno">  107</span>&#160;<span class="comment">//  data also gets cloned so that each genome has its own eval data (unlike the</span></div>
<div class="line"><a name="l00108"></a><span class="lineno">  108</span>&#160;<span class="comment">//  user data pointer described next which is shared by all genomes).</span></div>
<div class="line"><a name="l00109"></a><span class="lineno">  109</span>&#160;<span class="comment">//</span></div>
<div class="line"><a name="l00110"></a><span class="lineno">  110</span>&#160;<span class="comment">//userData</span></div>
<div class="line"><a name="l00111"></a><span class="lineno">  111</span>&#160;<span class="comment">//  The userData member is used to provide all genomes access to the same user</span></div>
<div class="line"><a name="l00112"></a><span class="lineno">  112</span>&#160;<span class="comment">//  data.  This can be a pointer to anything you want.  Any genome cloned from</span></div>
<div class="line"><a name="l00113"></a><span class="lineno">  113</span>&#160;<span class="comment">//  another will share the same userData as the original.  This means that all</span></div>
<div class="line"><a name="l00114"></a><span class="lineno">  114</span>&#160;<span class="comment">//  of the genomes in a population, for example, share the same userData.</span></div>
<div class="line"><a name="l00115"></a><span class="lineno">  115</span>&#160;<span class="comment">//</span></div>
<div class="line"><a name="l00116"></a><span class="lineno">  116</span>&#160;<span class="comment">//score</span></div>
<div class="line"><a name="l00117"></a><span class="lineno">  117</span>&#160;<span class="comment">//  Evaluate the &#39;performance&#39; of the genome using the objective function.</span></div>
<div class="line"><a name="l00118"></a><span class="lineno">  118</span>&#160;<span class="comment">//  The score is kept in the &#39;score&#39; member.  The &#39;evaluated&#39; member tells us</span></div>
<div class="line"><a name="l00119"></a><span class="lineno">  119</span>&#160;<span class="comment">//  whether or not we can trust the score.  Be sure to set/unset the &#39;evaluated&#39;</span></div>
<div class="line"><a name="l00120"></a><span class="lineno">  120</span>&#160;<span class="comment">//  member as appropriate (eg cross and mutate change the contents of the</span></div>
<div class="line"><a name="l00121"></a><span class="lineno">  121</span>&#160;<span class="comment">//  genome so they unset the &#39;evaluated&#39; flag).</span></div>
<div class="line"><a name="l00122"></a><span class="lineno">  122</span>&#160;<span class="comment">//    If there is no objective function, then simply return the score.  This</span></div>
<div class="line"><a name="l00123"></a><span class="lineno">  123</span>&#160;<span class="comment">//  allows us to use population-based evaluation methods (where the population</span></div>
<div class="line"><a name="l00124"></a><span class="lineno">  124</span>&#160;<span class="comment">//  method sets the score of each genome).</span></div>
<div class="line"><a name="l00125"></a><span class="lineno">  125</span>&#160;<span class="comment">*/</span></div>
<div class="line"><a name="l00126"></a><span class="lineno">  126</span>&#160;</div>
<div class="line"><a name="l00127"></a><span class="lineno"><a class="code" href="Chromosome_8h.html#a1ce5d72ed6a0cd74112f3edeef606ad9">  127</a></span>&#160;<span class="keyword">typedef</span> float (*<a class="code" href="Chromosome_8h.html#a1ce5d72ed6a0cd74112f3edeef606ad9">Evaluator</a>) (<a class="code" href="classChromosome.html">Chromosome</a> &amp;);</div>
<div class="line"><a name="l00128"></a><span class="lineno"><a class="code" href="Chromosome_8h.html#a5c82dfe9a11dd79e79203755e7b2c7e6">  128</a></span>&#160;<span class="keyword">typedef</span> void  (*<a class="code" href="Chromosome_8h.html#a5c82dfe9a11dd79e79203755e7b2c7e6">Initializer</a>) (<a class="code" href="classChromosome.html">Chromosome</a> &amp;);</div>
<div class="line"><a name="l00129"></a><span class="lineno"><a class="code" href="Chromosome_8h.html#a4adc8d5f73dcc5a13c24ca0ce968867d">  129</a></span>&#160;<span class="keyword">typedef</span> void  (*<a class="code" href="Chromosome_8h.html#a4adc8d5f73dcc5a13c24ca0ce968867d">Mutator</a>) (<a class="code" href="classChromosome.html">Chromosome</a> &amp;, float);</div>
<div class="line"><a name="l00130"></a><span class="lineno"><a class="code" href="Chromosome_8h.html#a305de7ffb4d0d863bd8d7e9598700b17">  130</a></span>&#160;<span class="keyword">typedef</span> int   (*<a class="code" href="Chromosome_8h.html#a305de7ffb4d0d863bd8d7e9598700b17">Comparator</a>) (<span class="keyword">const</span> <a class="code" href="classChromosome.html">Chromosome</a>&amp;, <span class="keyword">const</span> <a class="code" href="classChromosome.html">Chromosome</a>&amp;);</div>
<div class="line"><a name="l00131"></a><span class="lineno"><a class="code" href="Chromosome_8h.html#a1a8e042b027b0c262cfea265c4736f81">  131</a></span>&#160;<span class="keyword">typedef</span> void  (*<a class="code" href="Chromosome_8h.html#a1a8e042b027b0c262cfea265c4736f81">SexualCrossover</a>) (<span class="keyword">const</span> <a class="code" href="classChromosome.html">Chromosome</a>&amp;, <span class="keyword">const</span> <a class="code" href="classChromosome.html">Chromosome</a>&amp;,</div>
<div class="line"><a name="l00132"></a><span class="lineno">  132</span>&#160;                                  <a class="code" href="classChromosome.html">Chromosome</a>*, <a class="code" href="classChromosome.html">Chromosome</a>*);</div>
<div class="line"><a name="l00133"></a><span class="lineno"><a class="code" href="Chromosome_8h.html#ab2005e384577a3b12bb81f20438b8f7c">  133</a></span>&#160;<span class="keyword">typedef</span> void  (*<a class="code" href="Chromosome_8h.html#ab2005e384577a3b12bb81f20438b8f7c">AsexualCrossover</a>) (<span class="keyword">const</span> <a class="code" href="classChromosome.html">Chromosome</a>&amp;, <a class="code" href="classChromosome.html">Chromosome</a>*);</div>
<div class="line"><a name="l00134"></a><span class="lineno">  134</span>&#160;</div>
<div class="line"><a name="l00135"></a><span class="lineno"><a class="code" href="classChromosome.html">  135</a></span>&#160;<span class="keyword">class </span><a class="code" href="classChromosome.html">Chromosome</a> {</div>
<div class="line"><a name="l00136"></a><span class="lineno">  136</span>&#160;<span class="keyword">public</span>:</div>
<div class="line"><a name="l00137"></a><span class="lineno"><a class="code" href="classChromosome.html#a7eb065b8aa3799e0cca4608e2d3606b7">  137</a></span>&#160;  <a class="code" href="classChromosome.html#a7eb065b8aa3799e0cca4608e2d3606b7">Chromosome</a>() {}</div>
<div class="line"><a name="l00138"></a><span class="lineno">  138</span>&#160;  <span class="comment">//Chromosome(const Chromosome&amp; c) { copyFrom(c); }</span></div>
<div class="line"><a name="l00139"></a><span class="lineno"><a class="code" href="classChromosome.html#a82760daad33f1ba570dacb465fb66010">  139</a></span>&#160;  <span class="keyword">virtual</span> <a class="code" href="classChromosome.html#a82760daad33f1ba570dacb465fb66010">~Chromosome</a>() {}</div>
<div class="line"><a name="l00140"></a><span class="lineno">  140</span>&#160;</div>
<div class="line"><a name="l00141"></a><span class="lineno"><a class="code" href="classChromosome.html#ae9386312fb8117d869058c038c2b5460">  141</a></span>&#160;  <span class="keyword">virtual</span> <span class="keywordtype">void</span> <a class="code" href="classChromosome.html#ae9386312fb8117d869058c038c2b5460">copyFrom</a>(<span class="keyword">const</span> <a class="code" href="classChromosome.html">Chromosome</a>&amp; c) { (*this) = c; }</div>
<div class="line"><a name="l00142"></a><span class="lineno">  142</span>&#160;</div>
<div class="line"><a name="l00143"></a><span class="lineno"><a class="code" href="classChromosome.html#a1071ca1c0e3c43c900c1952b0a20309a">  143</a></span>&#160;  <span class="keyword">virtual</span> <span class="keywordtype">void</span> <a class="code" href="classChromosome.html#a1071ca1c0e3c43c900c1952b0a20309a">init</a>(){}</div>
<div class="line"><a name="l00144"></a><span class="lineno"><a class="code" href="classChromosome.html#a2a67c4966b4d05a19f851d850dbcfd95">  144</a></span>&#160;  <span class="keyword">virtual</span> <span class="keywordtype">void</span> <a class="code" href="classChromosome.html#a2a67c4966b4d05a19f851d850dbcfd95">mutate</a>(<span class="keywordtype">float</span> p){}</div>
<div class="line"><a name="l00145"></a><span class="lineno">  145</span>&#160;  <span class="keyword">virtual</span> <span class="keywordtype">bool</span> <a class="code" href="classChromosome.html#adeaa38fcbc4b70355112290e963c5c77">equals</a>(<span class="keyword">const</span> <a class="code" href="classChromosome.html">Chromosome</a>&amp; g) <span class="keyword">const</span> = 0;</div>
<div class="line"><a name="l00146"></a><span class="lineno">  146</span>&#160;<span class="comment">//  virtual float evaluate() { return 0; }</span></div>
<div class="line"><a name="l00147"></a><span class="lineno">  147</span>&#160;</div>
<div class="line"><a name="l00148"></a><span class="lineno">  148</span>&#160;<span class="comment">//  float evaluate() const { return (*evaluator)(*this); }</span></div>
<div class="line"><a name="l00151"></a><span class="lineno">  151</span>&#160;<span class="comment"></span><span class="comment">//</span></div>
<div class="line"><a name="l00152"></a><span class="lineno">  152</span>&#160;<span class="comment">//  void init() { (*initializer)(*this); }</span></div>
<div class="line"><a name="l00155"></a><span class="lineno">  155</span>&#160;<span class="comment"></span><span class="comment">//</span></div>
<div class="line"><a name="l00156"></a><span class="lineno">  156</span>&#160;<span class="comment">//  void mutate(float p) { (*mutator)(*this,p); }</span></div>
<div class="line"><a name="l00159"></a><span class="lineno">  159</span>&#160;<span class="comment"></span><span class="comment">//</span></div>
<div class="line"><a name="l00160"></a><span class="lineno">  160</span>&#160;<span class="comment">//  float compare(const Chromosome&amp; g) const { return (*comparator)(*this, g); }</span></div>
<div class="line"><a name="l00163"></a><span class="lineno">  163</span>&#160;<span class="comment"></span></div>
<div class="line"><a name="l00164"></a><span class="lineno">  164</span>&#160;};</div>
<div class="line"><a name="l00165"></a><span class="lineno">  165</span>&#160;</div>
<div class="line"><a name="l00166"></a><span class="lineno">  166</span>&#160;<span class="preprocessor">#endif </span><span class="comment">/* CHROMOSOME_H_ */</span><span class="preprocessor"></span></div>
</div><!-- fragment --></div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated on Fri Oct 11 2013 18:57:48 for Qualia by &#160;<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/>
</a> 1.8.3.1
</small></address>
</body>
</html>
